toolkit/tools/internal/safechroot/safechroot.go

```diff
@@ -689,3 +700,66 @@ func (c *Chroot) GetMountPoints() []*MountPoint {
 	mountPoints := append([]*MountPoint(nil), c.mountPoints...)
 	return mountPoints
 }
+
+// stopGPGComponents stops gpg-agent and keyboxd if they are running inside the chroot.
+//
+// A GPG agent may have been started while the chroot was in use. Newer versions of "gnupg2" will also start keyboxd.
+// E.g. when installing the azurelinux-repos-shared package, a GPG import occurs. This starts the gpg-agent process inside the chroot.
+// To be able to cleanly exit the setup chroot, we must stop it.
+func (c *Chroot) stopGPGComponents() (err error) {
+	_, err = exec.LookPath("gpgconf")
+	if err != nil {
+		// gpgconf is not installed, so gpg-agent is not running.
+		logger.Log.Debugf("gpgconf is not installed, so gpg-agent is not running: %s", err)
+		return
+	}
+
+	err = c.UnsafeRun(func() (err error) {
+		// Split --list-components stdout into a list of name tags, one for each component
+		components, err := c.listGPGComponents()
+		// List of components to kill. The names must be verbatim identical to the name tag that is used by `gpgconf`
+		componentsToKill := []string{"gpg-agent", "keyboxd"}
+		for _, component := range componentsToKill {
+			// Check if the line contains a component to kill
+			if components[component] {
+				logger.Log.Debugf("Found %s running inside chroot. Stopping it.", component)
+				_, stderr, err := shell.Execute("gpgconf", "--kill", component)
+				if err != nil {
+					err = fmt.Errorf("failed to stop GPG component (%s):\nerr: %w\nstderr: %s", component, err, stderr)
+					return err
+				}
+			}
+		}
+
+		return
+	})
+
+	return
+}
+
+func (c *Chroot) listGPGComponents() (components map[string]bool, err error) {
+	stdout, stderr, err := shell.Execute("gpgconf", "--list-components")
+
+	if err != nil {
+		err = fmt.Errorf("failed to list GPG components.\nerr: %w\nstderr: %s", err, stderr)
+		return
+	}
+
+	logger.Log.Debugf("gpgconf --list-components output:\n%s", stdout)
+
+	reader := strings.NewReader(stdout)
+	scanner := bufio.NewScanner(reader)
+
+	components = make(map[string]bool)
+	for scanner.Scan() {
+		line := scanner.Text()
+		components[strings.Split(line, ":")[0]] = true
+	}
+
+	err = scanner.Err()
+	if err != nil {
+		err = fmt.Errorf("error parsing gpgconf --list-components output: %w", err)
+	}
```

Comment:
I think the line parsing you had earlier was fine. I was probably confusing when I tried to explain my main concern: it was about the decreased readability of multiple nested indentations, the logic itself worked fine. And I did suggest a regex but that was just a dumb suggestion on my end, when I read more closely what we need to do here in terms of parsing.

Anyway, what do you think about this:
```suggestion
	components = make(map[string]bool)
	for _, line := range strings.Split(stdout, "\n") {
		components[strings.Split(line, ":")[0]] = true
	}
```